packages/core/echo/echo-db/src/query/query.ts

```diff
@@ -120,43 +102,40 @@ export type QuerySubscriptionOptions = {
   fire?: boolean;
 };
 
+export type QueryRunOptions = {
+  timeout?: number;
+};
+
 /**
  * Predicate based query.
  */
 export class Query<T extends {} = any> {
   private readonly _filter: Filter;
-  private readonly _sources = new Set<QuerySource>();
   private readonly _signal = compositeRuntime.createSignal();
   private readonly _event = new Event<Query<T>>();
+  private readonly _diagnostic: QueryDiagnostic;
 
-  private _runningCtx: Context | null = null;
-  private _resultCache: QueryResult<T>[] | undefined = undefined;
-  private _objectCache: EchoReactiveObject<T>[] | undefined = undefined;
+  private _isActive = false;
+  private _resultCache?: QueryResult<T>[] = undefined;
+  private _objectCache?: T[] = undefined;
   private _subscribers: number = 0;
 
-  private readonly _diagnostic: QueryDiagnostic;
-
   constructor(
     private readonly _queryContext: QueryContext,
     filter: Filter,
   ) {
     this._filter = filter;
 
-    this._queryContext.sources.forEach((s) => this._sources.add(s));
-    this._queryContext.added.on((source) => {
-      if (this._sources.has(source)) {
-        return;
-      }
-      this._sources.add(source);
-      if (this._runningCtx != null) {
-        this._subscribeToSourceUpdates(this._runningCtx, source);
-      }
-    });
-
-    this._queryContext.removed.on((source) => {
-      source.close();
-      this._sources.delete(source);
+    this._queryContext.changed.on(() => {
+      this._resultCache = undefined;
+      this._objectCache = undefined;
+      // Clear `prohibitSignalActions` to allow the signal to be emitted.
+      compositeRuntime.untracked(() => {
+        this._event.emit(this);
+        this._signal.notifyWrite();
+      });
     });
+    this._queryContext.update(filter);
```

Comment:
does it maybe have sense to move the filter to query context as well (so that it's not passed from query, seems to just complicated the query context API)